File: EasyClick.Web.Mini/JS源码/Pagination.js

Mini.ui.Pagination = function (options) {
    /// <summary>分页组件</summary>

    var defaults = {
        id: '',
        itemTotal: 500,
        urlFormat: "pageIndex={0}",
        rowCount: 10,
        buttonCount: 10,
        curPage: 1,
        pageCount: 0,
        command: '',
        buttonText: { first: "First", prev: "Prev", next: "Next", last: "Last" }
    };

    var m_Params = {
        id: '',
        itemTotal: 500,
        urlFormat: "pageIndex_{0}",
        rowCount: 10,
        buttonCount: 10,
        curPage: 0,
        pageCount: 0,
        command: '',
        buttonText: { first: "First", prev: "Prev", next: "Next", last: "Last" }
    };

    var m_Box;

    var m_StartPageIndex = 0;
    var m_EndPageIndex = 0;

    var m_ClickEvent;

    this.click = function (c) {
        var items = m_Box.children("a");

        $(items).click(c);

        m_ClickEvent = c;

    };

    function init(options) {
        /// <summary>初始化</summary>

        m_Params = $.extend(defaults, options);

        m_Box = $("#" + m_Params.id);

        if (m_Box.length == 0) {
            alert("Pagination.js \nfunction init(options) 初始化错误,不存在 html 对象'" + m_Params.id + "'");
        }
    }

    function getFirstUrl() {
        /// <summary>获取"首页"的超链接</summary>

        return $.format(m_Params.urlFormat, "0");
    }

    function getPrevUrl() {
        /// <summary>获取"上一页"的链接</summary>
        var p = m_Params.curPage - 1;

        if (p < 0) { p = 0; }

        return $.format(m_Params.urlFormat, p);
    }

    function getNextUrl() {
        /// <summary>获取"下一页"的链接</summary>

        var p = m_Params.curPage + 1;

        if (p >= m_Params.pageCount) {
            p = m_Params.pageCount - 1;
        }

        return $.format(m_Params.urlFormat, p);
    }

    function getLastUrl() {
        /// <summary>获取"尾页"的链接</summary>

        return $.format(m_Params.urlFormat, m_Params.pageCount - 1);
    }

    this.reset = function (options) {
        /// <summary>重新设置</summary>

        for (var i in options) {
            if (m_Params[i] != undefined) {
                m_Params[i] = options[i];
            }
        }


        var inputBox = $("#" + m_Params.id + "_CurPIndex");
        $(inputBox).val(m_Params.curPage);

        var m = m_Params.itemTotal % m_Params.rowCount;

        m_Params.pageCount = (m_Params.itemTotal - m) / m_Params.rowCount;

        if (m > 0) {
            m_Params.pageCount++;
        }

        m_StartPageIndex = m_Params.curPage - 4;
        m_EndPageIndex = m_Params.curPage + 5;

        if (m_StartPageIndex < 0) {
            m_StartPageIndex = 0;

            //            m_EndPageIndex += 4 - m_Params.curPage;
        }

        if (m_EndPageIndex >= m_Params.pageCount) {
            m_EndPageIndex = m_Params.pageCount - 1;
        }

        //重新构造分页的所有按钮
        m_Box.html("");

        var btnFormat = "<a href='{0}' command='{3}' commandParam='{2}' valid='false'>{1}</a>";

        var btnTxt = m_Params.buttonText;

        if (m_Params.curPage > 0) {
            m_Box.append($.format(btnFormat, getFirstUrl(), btnTxt.first, 0, m_Params.command));
        }
        else {
            m_Box.append($.format("<span class='disabled'>{0}</span>", btnTxt.first));
        }

        if (m_Params.curPage > 0) {
            m_Box.append($.format(btnFormat, getPrevUrl(), btnTxt.prev, m_Params.curPage - 1, m_Params.command));
        }
        else {
            m_Box.append($.format("<span class='disabled'>{0}</span>", btnTxt.prev));
        }

        for (var i = m_StartPageIndex; i <= m_EndPageIndex; i++) {

            var url = $.format(m_Params.urlFormat, i);
            if (i == m_Params.curPage) {
                m_Box.append($.format("<span class='current'>{0}</span>", i + 1));
            }
            else {
                m_Box.append($.format(btnFormat, url, i + 1, i, m_Params.command));
            }

        }

        if (m_Params.curPage < m_Params.pageCount - 1) {
            m_Box.append($.format(btnFormat, getNextUrl(), btnTxt.next, m_Params.curPage + 1, m_Params.command));
            m_Box.append($.format(btnFormat, getLastUrl(), btnTxt.last, m_Params.pageCount - 1, m_Params.command));
        }
        else {
            m_Box.append($.format("<span class='disabled'>{0}</span>", btnTxt.next));
            m_Box.append($.format("<span class='disabled'>{0}</span>", btnTxt.last));
        }

        //给各个按钮加上事件
        if (m_ClickEvent) {
            var items = m_Box.children("a");
            $(items).click(m_ClickEvent);
        }

        var startRowIndex = m_Params.curPage * m_Params.rowCount;
        var endRowIndex = startRowIndex + m_Params.rowCount;

        if (endRowIndex > m_Params.itemTotal) {
            endRowIndex = m_Params.itemTotal;
        }

        m_Box.append($.format("当前记录 {0}--{1} 条,共 {2} 条记录", startRowIndex + 1, endRowIndex, m_Params.itemTotal));
    }

    init(options);
};

File: EasyClick.Web.Mini/Pagination.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Web.UI;
using System.ComponentModel;
using EasyClick.Web.Mini.Utility;

namespace EasyClick.Web.Mini
{
    /// <summary>
    /// 分页组件
    /// </summary>
    [ToolboxData("<{0}:Pagination runat=\"server\" className=\"\" />")]
    public class Pagination:Control,IClientIDMode,IMiniControl
    {
        /// <summary>
        /// 分页组件
        /// </summary>
        public Pagination()
        {

        }

        #region 字段

        string m_Command = "GoPage";

        /// <summary>
        /// 样式名称
        /// </summary>
        string m_ClassName = "sabrosus";

        /// <summary>
        /// 每页显示的记录数量
        /// </summary>
        int m_RowCount = 20;

        
        string m_UrlFormat = "PageIndex={0}";

        /// <summary>
        /// 显示的页数量
        /// </summary>
        int m_ButtonCount = 10;
        
        /// <summary>
        /// 当前页码
        /// </summary>
        int m_CurPage = 0;

        /// <summary>
        /// 记录总数量
        /// </summary>
        int m_ItemTotal = 0;

        string m_ButtonFirstText = "首页";

        string m_ButtonPrevText = "上一页";

        string m_ButtonNextText = "下一页";
        string m_ButtonLastText = "尾页";

        /// <summary>
        ///
        /// </summary>
        bool m_EcDesignMode = false;

        /// <summary>
        /// 设置为 AJAX
        /// </summary>
        bool m_IsAjax = false;

        CellAlign m_Align = CellAlign.Center;

        #endregion

        #region 按钮的文字

        [DefaultValue("首页")]
        [Description("首页的文字")]
        public string ButtonFistText
        {
            get { return m_ButtonFirstText; }
            set { m_ButtonFirstText = value; }
        }

        [DefaultValue("上一页")]
        [Description("上一页的文字")]
        public string ButtonPrevText
        {
            get { return m_ButtonPrevText; }
            set { m_ButtonPrevText = value; }
        }


        [DefaultValue("下一页")]
        [Description("下一页的文字")]
        public string ButtonNextText
        {
            get { return m_ButtonNextText; }
            set { m_ButtonNextText = value; }
        }

        [DefaultValue("尾页")]
        [Description("尾页的文字")]
        public string ButtonLastText
        {
            get { return m_ButtonLastText; }
            set { m_ButtonLastText = value; }
        }
        #endregion


        #region 属性

        [DefaultValue(CellAlign.Center)]
        public CellAlign Align
        {
            get { return m_Align; }
            set { m_Align = value; }
        }

        /// <summary>
        /// 设置 Ajax 模式
        /// </summary>
        [DefaultValue(false)]
        [Description("设置 Ajax 模式")]
        public bool IsAjax
        {
            get { return m_IsAjax; }
            set { m_IsAjax = value; }
        }

        /// <summary>
        /// 命令名称
        /// </summary>
        [DefaultValue("")]
        [Description("命令名称")]
        public string Command
        {
            get { return m_Command; }
            set { m_Command = value; }
        }

        /// <summary>
        /// 样式文件
        /// </summary>
        [DefaultValue("")]
        [Description("当前页码")]
        public string ClassName
        {
            get { return m_ClassName; }
            set { m_ClassName = value; }
        }

        /// <summary>
        /// 每页显示的记录数量
        /// </summary>
        [DefaultValue(0)]
        [Description("每页显示的记录数量")]
        public int RowCount
        {
            get { return m_RowCount; }
            set { m_RowCount = value; }
        }

        /// <summary>
        /// 按钮的超链接
        /// </summary>
        [DefaultValue("")]
        [Description("按钮的超链接")]
        public string UrlFormat
        {
            get { return m_UrlFormat; }
            set { m_UrlFormat = value; }
        }

        /// <summary>
        /// 按钮的数量
        /// </summary>
        [DefaultValue(0)]
        [Description("按钮的数量")]
        public int ButtonCount
        {
            get { return m_ButtonCount; }
            set { m_ButtonCount = value; }
        }

        /// <summary>
        /// 当前页码
        /// </summary>
        [DefaultValue(0)]
        [Description("当前页码")]
        public int CurPage
        {
            get { return m_CurPage; }
            set { m_CurPage = value; }
        }

        /// <summary>
        /// 记录的总数量
        /// </summary>
        [DefaultValue(0)]
        [Description("记录的总数量")]
        public int ItemTotal
        {
            get { return m_ItemTotal; }
            set { m_ItemTotal = value; }
        }

        [DefaultValue(false)]
        public bool EcDesignMode
        {
            get { return m_EcDesignMode; }
            set { m_EcDesignMode = value; }
        }

        #endregion


        ClientIDMode m_ClientIDMode = ClientIDMode.AutoID;

        [DefaultValue(ClientIDMode.AutoID)]
        public ClientIDMode ClientIDMode
        {
            get { return m_ClientIDMode; }
            set { m_ClientIDMode = value; }
        }



        public string GetClientID()
        {
            string cId;

            switch (m_ClientIDMode)
            {
                case ClientIDMode.Static:
                    cId = this.ID;
                    break;
                default:
                    cId = this.ClientID;
                    break;
            }

            return cId;
        }


        public void Reset()
        {
            MiniScript.Add("{0}.reset({{'itemTotal':{1},'curPage':{2},'urlFormat':'{3}','buttonCount':{4},'rowCount':{5},'command':'{6}' }});", 
                GetClientID(),
                this.ItemTotal, this.CurPage,this.UrlFormat, this.ButtonCount, this.RowCount,this.Command);

            MiniScript.Add("$('#{0}_CurPIndex').val('{1}');", this.GetClientID(), this.CurPage);
        }

        bool m_Visible = true;

        [DefaultValue(true)]
        public override bool Visible
        {
            get { return m_Visible; }
            set { m_Visible = value; }
        }


        protected override void Render(HtmlTextWriter writer)
        {

            if (this.DesignMode || this.EcDesignMode)
            {
                if (!m_Visible)
                {
                    return;
                }

                writer.Write("<div class='{0}' ",this.ClassName);
                writer.Write(">");

                writer.Write("<a >{0}</a>", m_ButtonFirstText);
                writer.Write("<a >{0}</a>", m_ButtonPrevText);

                writer.Write("<a >1</a>");
                writer.Write("<a >2</a>");
                writer.Write("<a >3</a>");
                writer.Write("<a >4</a>");


                writer.Write("<a >{0}</a>", m_ButtonNextText);
                writer.Write("<a >{0}</a>", m_ButtonLastText);

                writer.Write("</div>");

                return;
            }


            writer.Write("<div id='{0}' class='{1}' ", this.GetClientID(), this.ClassName);
            if (!m_Visible)
            {
                writer.Write("style='display:none;' ");
            }
            writer.Write("></div>");

            writer.WriteLine("<input type='hidden' id='{0}_CurPIndex' name='{0}_CurPIndex' value='0' />", this.GetClientID());
            writer.WriteLine("<input type='hidden' id='{0}_RowCount' name='{0}_RowCount' value='{1}' />", this.GetClientID(),m_RowCount);


            writer.WriteLine("<script type='text/javascript'>");
            writer.WriteLine("var {0} = null;", this.GetClientID());

            string jsMode = null;
            if (System.Web.HttpContext.Current != null && System.Web.HttpContext.Current.Items.Contains("JS_MODE"))
            {
                jsMode = (string)System.Web.HttpContext.Current.Items["JS_MODE"];
            }

            if (jsMode == "InJs")
            {
                writer.WriteLine("In.ready('mi.Pagination',function(){");
            }
            else
            {
                writer.WriteLine("$(document).ready(function(){");
            }

            writer.WriteLine("    var c = new Mini.ui.Pagination({ id: '" + this.GetClientID() + "',allowSetupRowCount:true });");
            writer.WriteLine("    c.reset({");

            writer.WriteLine("        itemTotal:{0},", this.ItemTotal);
            writer.WriteLine("        curPage:{0},", this.CurPage);
            writer.WriteLine("        urlFormat:'{0}',", this.UrlFormat);
            writer.WriteLine("        buttonCount:{0},", this.ButtonCount);
            writer.WriteLine("        rowCount:{0},", RowCount);
            writer.WriteLine("        command:'{0}',", this.Command);
            writer.WriteLine("        buttonText:{{first:'{0}',prev:'{1}',next:'{2}',last:'{3}'}}", m_ButtonFirstText, m_ButtonPrevText, m_ButtonNextText, m_ButtonLastText);
            writer.WriteLine("    });");

            writer.WriteLine("    {0} = c;", this.GetClientID());

            if (m_IsAjax)
            {
                UserControl uc = FindWidget();

                if (uc != null)
                {
                    writer.WriteLine("  c.click(function (sender,e) {");

                    //writer.WriteLine("    if(e && e.owner && e.owner.getRowCount){");
                    //writer.WriteLine("        $('#{0}_RowCount').val( e.owner.getRowCount() );", GetClientID());
                    //writer.WriteLine("    }");

                    writer.WriteLine("    $('#{0}_CurPIndex').val( $(this).attr('commandParam') );", GetClientID());
                    writer.WriteLine("    {0}.submit(this);", uc.ClientID);
                    writer.WriteLine("    return false;");
                    writer.WriteLine("  });");
                }
            }

            writer.WriteLine("});");
            writer.WriteLine("</script>");

        }

        private UserControl FindWidget()
        {
            Control con = this.Parent;

            for (int i = 0; i < 9; i++)
            {
                if (con is UserControl)
                {
                    break;
                }

                if (con.Parent == null)
                {
                    break;
                }

                con = con.Parent;
            }

            return con as UserControl;
        }

        #region IMiniControl 成员

        public void LoadPostData()
        {
            string key = string.Format("{0}_CurPIndex", GetClientID());
            string vStr = this.Context.Request.Form[key];

            string rowCountId = string.Format("{0}_RowCount", GetClientID());
            string rowCountStr = this.Context.Request.Form[rowCountId];
            
            this.m_CurPage = StringUtility.ToInt(vStr);
            this.m_RowCount = StringUtility.ToInt(rowCountStr);
        }

        #endregion
    }
}
